### ISSUE TITLE
TagzApp OverlayTests failure troubleshooting

### ISSUE THREAD

**Issue by eric-vanartsdalen** [CZEMacLeod/C3D.Extensions.Playwright.AspNetCore#1]
Hi @CZEMacLeod, I'm trying to work an implementation to TagzApp for Moderation database.

This test (OverlayTests) is failing, and am having a hard time troubleshooting the cause. How would I configure the calls to allow for headless=false or tracing of each page's output?

Thanks in Advance for any insights and help.
Cheers,
Eric

See my repo for details:  
https://github.com/eric-vanartsdalen/TagzApp/blob/34_Moderation/src/TagzApp.WebTest/Tests/OverlayTests.cs

**Comment by CZEMacLeod** [comment]
@eric-vanartsdalen 
There are a couple of ways of doing this. I have comments in the MyFirstTests to show how to quickly do traces (although the full manual way also works - it is just an extension method and a [disposable object](https://github.com/CZEMacLeod/C3D.Extensions.Playwright.AspNetCore/blob/main/src/C3D/Extensions/Playwright/AspNetCore/Utilities/PlaywrightTrace.cs) that closes the trace on dispose to make it easier).

1. In the test you can do 

```cs

    await using var mainTrace = await mainPage.TraceAsync("Click Message On Waterfall", true, true, true);
    mainTrace.Show = C3D.Extensions.Playwright.AspNetCore.Utilities.PlaywrightTraceShow.OnClose;

    await using var overlayTrace = await mainPage.TraceAsync("Overlay shows message from waterfall", true, true, true);
    overlayTrace.Show = C3D.Extensions.Playwright.AspNetCore.Utilities.PlaywrightTraceShow.OnClose;

```

Don't check in the `Show = OnClose` as this will launch the show trace when in the container for tests etc. This is also why I have headless set to true by default. It would be possible to add a property to the config when running the tests in a container, and check for that, and by default show the browser, but that requires additional work on the consumer of the package to setup. It makes more sense to require a consumer to opt in to showing the browser etc.

2. You can edit the `PlaywrightFixture` class to override the default browser launch options to show the browser.
```cs
  protected override BrowserTypeLaunchOptions LaunchOptions
  {
    get
    {
      var options = base.LaunchOptions;
      options.Headless = false;
      return options;
    }
  }
```
Again - don't leave this in when checking in an update as it doesn't play well with running in a container or build server for CI tests.

3. You can launch a new browser instance and use that instead.
```cs
    await using var browser1 = await WebApp.CreateCustomPlaywrightBrowserPageAsync(browserOptions: (options) => { options.Headless = false; });
    await using var browser2 = await WebApp.CreateCustomPlaywrightBrowserPageAsync(browserOptions: (options) => { options.Headless = false; });
    var mainPage = browser1.Page;
    var overlayPage = browser2.Page;
```
This creates 2 browser instances, instead of 2 separate contexts in a single browser, and adds additional config to each browser instance to override the default `LaunchOptions` from the fixture.
The returned object is just a [disposable wrapper](https://github.com/CZEMacLeod/C3D.Extensions.Playwright.AspNetCore/blob/main/src/C3D/Extensions/Playwright/AspNetCore/Utilities/PlaywrightBrowserPage.cs) around the browser, and a page created in that browser, that does all the close/shutdown for the page and browser on dispose, so you don't have to deal with that stuff in the test. Note that you need to comment out the existing code that creates the 2 [context wrappers](https://github.com/CZEMacLeod/C3D.Extensions.Playwright.AspNetCore/blob/main/src/C3D/Extensions/Playwright/AspNetCore/Utilities/PlaywrightContextPage.cs) lines 20-24 which work the same way.

N.B. It appears that playwright will only show 1 trace at a time, so using `Show = OnClose` won't work as expected.
If you use `Show = OnCloseAndWait` it won't exit the test until the trace window is closed, and will show one trace then the other.

The code for the packages is available at [CZEMacLeod/C3D.Extensions.Playwright.AspNetCore](https://github.com/CZEMacLeod/C3D.Extensions.Playwright.AspNetCore). This contains both `C3D.Extensions.Playwright.AspNetCore` and `C3D.Extensions.Playwright.AspNetCore.Xunit`.
And if you need any further input on this, please feel free to ask here, or open an issue on the package repo if it is something specific to the generic fixtures.

There is some documentation in the NuGet package readme files:

- [C3D.Extensions.Playwright.AspNetCore](https://github.com/CZEMacLeod/C3D.Extensions.Playwright.AspNetCore/blob/main/src/C3D/Extensions/Playwright/AspNetCore/README.md)
- [C3D.Extensions.Playwright.AspNetCore.Xunit](https://github.com/CZEMacLeod/C3D.Extensions.Playwright.AspNetCore/blob/main/src/C3D/Extensions/Playwright/AspNetCore.Xunit/README.md)

But I'm open to PRs and help with documenting it better!

**Comment by CZEMacLeod** [comment]
Interesting idea here - and I wonder if it is stable enough to put it into the main package implementation:
In `CreateHost` for your fixture
```cs
builder.ConfigureAppConfiguration(config => config.AddUserSecrets<PlaywrightFixture>());
```
This grabs the usersecrets id from the test assembly (hence it can't be done in the base class).
Then we use the configuration from the host (augmented with some test secrets), to configure the tests.
```cs
  protected override BrowserTypeLaunchOptions LaunchOptions
  {
    get
    {
      var options = base.LaunchOptions;
      var config = Services.GetRequiredService<IConfiguration>();
      config.GetSection("Playwright:BrowserTypeLaunchOptions").Bind(options);
      return options;
    }
  }

  protected override BrowserNewContextOptions ContextOptions
  {
    get
    {
      var options = base.ContextOptions;
      var config = Services.GetRequiredService<IConfiguration>();
      config.GetSection("Playwright:BrowserNewContextOptions").Bind(options);
      return options;
    }
  }

  protected override BrowserNewPageOptions PageOptions
  {
    get
    {
      var options = base.PageOptions;
      var config = Services.GetRequiredService<IConfiguration>();
      config.GetSection("Playwright:BrowserNewPageOptions").Bind(options);
      return options;
    }
  }

  public override PlaywrightBrowserType BrowserType
  {
    get
    {
      var config = Services.GetRequiredService<IConfiguration>();
      return config.GetSection("Playwright").GetValue<PlaywrightBrowserType>("BrowserType", base.BrowserType);
    }
  }
```
Then in the user secrets for the test project
```json
{
	"Playwright": {
		"BrowserTypeLaunchOptions": {
			"Headless": false
		},
		"BrowserType": "Firefox"
	}
}
```
N.B. I'm only using firefox here for an example to test it and it is _not_ very stable.
When in the container/build server, there are no secrets, so it runs the tests headless as expected...
With a bit of work, this might be possible to extend to the trace mechanism too, to allow setting the `TracingStartOptions`.

Let me think on this and maybe open an issue in [CZEMacLeod/C3D.Extensions.Playwright.AspNetCore](https://github.com/CZEMacLeod/C3D.Extensions.Playwright.AspNetCore) to add this as a feature enhancement.

**Comment by CZEMacLeod** [comment]
I've opened a discussion #3 regarding adding the feature to load the playwright options from the host configuration, and automatically adding the test project's settings (appsettings.json and user secrets). If you have any feedback on the idea, let me know there.

**Comment by eric-vanartsdalen** [comment]
Thanks for the above... I used your point 3  launch a new browser instance with arguments browserOptions: (options) => { options.Headless = false; }...
In my attempt to extract a more generic service in the TagzApp, I broke the way the card click resulting in the overlay page getting the popup comms. So I'm not sure how to get that back working... time will tell. When I get more time, I'll try some more. Thanks Eric

**Comment by CZEMacLeod** [comment]
I have started an initial draft of the configuration from host/test/secrets settings in the [feature-configuration](https://github.com/CZEMacLeod/C3D.Extensions.Playwright.AspNetCore/tree/feature-configuration) branch of this project.

I haven't started adding tests for it yet, but I'm pretty happy with how it is turning out.
I've also made tracing a little easier to configure, and offer a disposable wrapper for the basic page (in addition to the context or custom browser wrappers). This means that the page will close at the end of the test automatically and you don't have to worry about using more resources than needed, especially on a build server.
